In [1]:
from pyspark.sql import SparkSession
from pyspark.sql.types import *
              
ticketSchema=StructType([StructField("Summons_Number",StringType(), True),
StructField("Plate_ID",StringType(), True),
StructField("Registration_State", StringType(), True),
StructField("Plate_Type", StringType(), True),
StructField("Issue_Date", StringType(), True),
StructField("Violation_Code", StringType(), True),
StructField("Vehicle_Body_Type", StringType(), True),
StructField("Vehicle_Make", StringType(), True),
StructField("Issuing_Agency", StringType(), True),
StructField("Street_Code1", StringType(), True),
StructField("Street_Code2", StringType(), True),
StructField("Street_Code3", StringType(), True),
StructField("Vehicle_Expiration_Date", StringType(), True),
StructField("Violation_Location", StringType(), True),
StructField("Violation_Precinct", StringType(), True),
StructField("Issuer_Precinct", StringType(), True),
StructField("Issuer_Code", StringType(), True),
StructField("Issuer_Command", StringType(), True),
StructField("Issuer_Squad", StringType(), True),
StructField("Violation_Time", StringType(), True),
StructField("Time_First_Observed", StringType(), True),
StructField("Violation_County", StringType(), True),
StructField("Violation_In_Front_Opposite", StringType(), True),
StructField("House_Number", StringType(), True),
StructField("Street_Name", StringType(), True),
StructField("Intersecting_Street", StringType(), True),
StructField("Date_First_Observed", StringType(), True),
StructField("Law_Section", StringType(), True),
StructField("Sub_Division", StringType(), True),
StructField("Violation_Legal_Code", StringType(), True),
StructField("Days_Parking_In_Effect", StringType(), True),
StructField("From_Hours_In_Effect", StringType(), True),
StructField("To_Hours_In_Effect", StringType(), True),
StructField("Vehicle_Color", StringType(), True),
StructField("Unregistered_Vehicle?", StringType(), True),
StructField("Vehicle_Year", StringType(), True),
StructField("Meter_Number", StringType(), True),
StructField("Feet_From_Curb", StringType(), True),
StructField("Violation_Post_Code", StringType(), True),
StructField("Violation_Description", StringType(), True),
StructField("No_Standing_or_Stopping_Violation", StringType(), True),
StructField("Hydrant_Violation", StringType(), True),
StructField("Double_Parking_Violation", StringType(), True),
StructField("Latitude", StringType(), True),
StructField("Longitude", StringType(), True),
StructField("Community_Board", StringType(), True),
StructField("Community_Council", StringType(), True),
StructField("Census_Tract", StringType(), True),
StructField("BIN", StringType(), True),
StructField("BBL", StringType(), True),
StructField("NTA", StringType(), True)
])

import re
import usaddress
from pyspark.sql import functions as F

directions={"se":"Southeast","ne":"Northeast","sw":"Southwest","nw":"Northwest","e":"East","w":"West","n":"North","s":"South",
           "southeast":"Southeast","northeast":"Northeast","southwest":"Southwest","northwest":"Northwest","east":"East","west":"West","north":"North","south":"South"}

streetTypes={"st":"Street","ave":"Avenue","blvd":"Boulevard","cv":"Cove","ct":"Court","cr":"Circle","ln":"Lane","dr":"Drive",
             "street":"Street","avenue":"Avenue","boulevard":"Boulevard","cove":"Cove","court":"Court","circle":"Circle","lane":"Lane","drive":"Drive"}

@F.udf(returnType=BooleanType())
def has_no_special_chars(value):
  if(re.search("\W+",str(value))):
    return False
  else:
    return True
@F.udf(returnType=BooleanType())
def has_special_chars(value):
  if(re.search("\W+",str(value))):
    return True
  else:
    return False
  
@F.udf(returnType=BooleanType())
def is_valid_date(value):
  if(re.match("^\d\d/\d\d/\d\d\d\d$",str(value))):
    return True
  else:
    return False
@F.udf(returnType=StringType())
def parse(houseNumber,streetName):
  addressStr=""
  if(houseNumber):
    addressStr=houseNumber+" "+streetName
  else:
    addressStr=streetName
  address=usaddress.parse(addressStr)
  addressStr=""
  for key in address:
    print(key)
    if(key[1]=="StreetNamePreDirectional" or key[1]=="StreetNamePostDirectional" ):
      addressStr+=" "+directions.get(key[0].lower(),key[0].lower())
    elif(key[1]=="StreetName"):
      try:
        strName=int(key[0])
        digit=strName%10
        if(digit in [0,4,5,6,7,8,9]):
          addressStr+=" "+key[0]+"th"
        elif(digit==1):
          addressStr+=" "+key[0]+"st"
        elif(digit==2):
          addressStr+=" "+key[0]+"nd"
        elif(digit==3):
          addressStr+=" "+key[0]+"rd"
      except:
        addressStr+=" "+key[0].lower()
    elif(key[1]=="StreetNamePostType"):
      addressStr+=" "+streetTypes.get(key[0].lower(),key[0].lower())
    else:
      addressStr+=" "+key[0].lower()
  return addressStr 



In [2]:
df=spark.read.csv("s3a://<key>:<secret_key>@nycpkviolations/Parking_Violations_Issued_-_Fiscal_Year_2015.csv",header=True,schema=ticketSchema)

In [3]:
for key in df.columns:
  print("Not Null count for "+key+" "+str(df.dropna(how="any",subset=key).count()))

Not Null count for Summons_Number 11809233
Not Null count for Plate_ID 11809231
Not Null count for Registration_State 11809233
Not Null count for Plate_Type 11809233
Not Null count for Issue_Date 11809233
Not Null count for Violation_Code 11809233
Not Null count for Vehicle_Body_Type 11763499
Not Null count for Vehicle_Make 11733723
Not Null count for Issuing_Agency 11809233
Not Null count for Street_Code1 11809233
Not Null count for Street_Code2 11809233
Not Null count for Street_Code3 11809233
Not Null count for Vehicle_Expiration_Date 11809233
Not Null count for Violation_Location 10010063
Not Null count for Violation_Precinct 11809233
Not Null count for Issuer_Precinct 11809233
Not Null count for Issuer_Code 11809233
Not Null count for Issuer_Command 10025775
Not Null count for Issuer_Squad 10025598
Not Null count for Violation_Time 11807518
Not Null count for Time_First_Observed 1227444
Not Null count for Violation_County 9969499
Not Null count for Violation_In_Front_Opposite 9927946
Not Null count for House_Number 9818474
Not Null count for Street_Name 11803178
Not Null count for Intersecting_Street 3243546
Not Null count for Date_First_Observed 11809232
Not Null count for Law_Section 11809232
Not Null count for Sub_Division 11808821
Not Null count for Violation_Legal_Code 1785147
Not Null count for Days_Parking_In_Effect 8970678
Not Null count for From_Hours_In_Effect 6622631
Not Null count for To_Hours_In_Effect 6622631
Not Null count for Vehicle_Color 11676099
Not Null count for Unregistered_Vehicle? 1380320
Not Null count for Vehicle_Year 11809230
Not Null count for Meter_Number 2263444
Not Null count for Feet_From_Curb 11809230
Not Null count for Violation_Post_Code 8647034
Not Null count for Violation_Description 10428878
Not Null count for No_Standing_or_Stopping_Violation 1
Not Null count for Hydrant_Violation 1
Not Null count for Double_Parking_Violation 1
Not Null count for Latitude 0
Not Null count for Longitude 0
Not Null count for Community_Board 0
Not Null count for Community_Council 0
Not Null count for Census_Tract 0
Not Null count for BIN 0
Not Null count for BBL 0
Not Null count for NTA 0

from the above counts, it is pretty clear that we can drop the columns with 0 counts from the final table.
Also sparks schema inference was not working properly, so I have defined all columns as string. I will convert them to the right type using dataframe options.

Going, from the first look of the data, we could probably have the following dimensions for this data. The fact for this data would be the actual summons or the ticket itself. Hence, the primary key for thid dataset is the "Summons Number" . The logical dimensions for such data sets would

1. The vehicle for which a ticket was provided
2. The violation code and the corresponding violation details.
3. The issuer of the ticket.
4. The location of the violation or the meter number.

We will do a quick analysis of these different dimension options and see if these can actually be stored as references.

Lets look at the vehicle dimension. The Tag or the Plate ID of the vehical would logically be the key for this particular dimension.

In [6]:
display(df.where(has_special_chars(df["Plate_ID"])))

Summons_Number Plate_ID Registration_State Plate_Type Issue_Date Violation_Code Vehicle_Body_Type Vehicle_Make Issuing_Agency Street_Code1 Street_Code2 Street_Code3 Vehicle_Expiration_Date Violation_Location Violation_Precinct Issuer_Precinct Issuer_Code Issuer_Command Issuer_Squad Violation_Time Time_First_Observed Violation_County Violation_In_Front_Opposite House_Number Street_Name Intersecting_Street Date_First_Observed Law_Section Sub_Division Violation_Legal_Code Days_Parking_In_Effect From_Hours_In_Effect To_Hours_In_Effect Vehicle_Color Unregistered_Vehicle? Vehicle_Year Meter_Number Feet_From_Curb Violation_Post_Code Violation_Description No_Standing_or_Stopping_Violation Hydrant_Violation Double_Parking_Violation Latitude Longitude Community_Board Community_Council Census_Tract BIN BBL NTA 7710757820 3.46E+87 MD PAS 09/05/2014 14 DELV HINO T 34850 10410 10510 01/01/20140088 12:00:00 PM 0018 18 18 349867 T503 A1 0617A null NY O 33 W 54th St null 01/05/0001 12:00:00 PM 408 c null YYYYYYY null null WHITE null 0 null 0 17 7 14-No Standing null null null null null null null null null null null 8013466176 N/S NY PAS 01/17/2015 74 4DSD ME/BE T 60810 40404 40404 01/01/88888888 12:00:00 PM 105 105 105 355712 T402 J 0735A null Q F 216-02 Rockaway Blvd null 01/05/0001 12:00:00 PM 408 j2 null null null null TAN null 0 null 0 15 4 74-Missing Display Plate null null null null null null null null null null null 8013466176 N/S NY PAS 01/17/2015 74 4DSD ME/BE T 60810 40404 40404 01/01/88888888 12:00:00 PM 105 105 105 355712 T402 J 0735A null Q F 216-02 Rockaway Blvd null 01/05/0001 12:00:00 PM 408 j2 null null null null TAN null 0 null 0 15 4 74-Missing Display Plate null null null null null null null null null null null 7441832494 ] NY PAS 02/06/2015 14 null null T 68190 40404 40404 01/01/20150101 12:00:00 PM 103 103 103 340127 T402 L 0747A null Q O null Woodhull Ave null 01/05/0001 12:00:00 PM 408 c null Y 0700A 1000A null null 0 null 0 B 42 14-No Standing null null null null null null null null null null null 7498841137 N/S NY MOT 10/30/2014 51 MCY YAMAH T 9890 9840 59990 01/01/88888888 12:00:00 PM 108 108 108 345395 T740 B 1011A null Q F 43-09 43rd St null 01/05/0001 12:00:00 PM 408 e3 null null null null BLACK null 0 null 0 01 6 51-Sidewalk null null null null null null null null null null null 7639987477 N/S NY PAS 08/18/2014 51 SUBN ME/BE T 25220 0 0 01/01/88888888 12:00:00 PM 103 103 103 356353 T402 H 0935A null Q I N 187th Pl 40ft W/of Jamaica Av 01/05/0001 12:00:00 PM 408 e3 null null null null SILVE null 0 null 0 10 4 51-Sidewalk null null null null null null null null null null null 7061894969 N/S NY 999 06/26/2015 14 SUBN HONDA T 34550 11720 97500 01/01/88888888 12:00:00 PM 0014 14 14 346109 T106 A 0813A null NY F 528 W 39th St null 01/05/0001 12:00:00 PM 408 c null YYYYYYY 0800A 1000P BLUE null 0 null 0 I 17 14-No Standing null null null null null null null null null null null 7095875560 XAVH72] NJ PAS 09/12/2014 46 VAN FORD T 18850 10410 25390 01/01/88888888 12:00:00 PM 0019 19 19 352950 T503 X1 0138P null NY F 24 E 94th St null 01/05/0001 12:00:00 PM 408 f1 null null null null WHITE null 0 null 0 01 7 46A-Double Parking (Non-COM) null null null null null null null null null null null 7082173106 AEE,5519 NY TRA 11/06/2014 21 LTRL INTER T 68930 48830 27530 01/01/88888888 12:00:00 PM 0084 84 84 347663 T301 Y 0916A null K O 120 Pacific St null 01/05/0001 12:00:00 PM 408 d1 null Y 0900A 1030A WHITE null 0 null 0 01 3 21-No Parking (street clean) null null null null null null null null null null null 7537475507 626)7JM NY COM 09/16/2014 38 DELV INTER T 63030 89730 46630 01/01/88888888 12:00:00 PM 0084 84 84 361034 T301 P 0251P null K F 475 Myrtle Ave null 01/05/0001 12:00:00 PM 408 h1 null Y 0830A 0700P BROWN null 1995 null 0 06 3 38-Failure to Display Muni Rec null null null null null null null null null null null 7355464220 HVP$102 PA PAS 09/05/2014 38 4DSD CHRYS T 25390 18450 18470 01/01/20150588 12:00:00 PM 0019 19 19 3

seeing the output above, we can see that the Plate_ID field has special characters in them, which cannot be tag numbers. But, overall, there the count of data where the Plate_ID is bad is very low compared to where the data is good. We can still use this as one of our dimensions.

In [8]:
#remove data with invalid plateID
print(df.where(has_no_special_chars(df["Plate_ID"])).count())
print(df.count())

11803733
11809233

so roughly 5500 records have bad plate ID. For the purpose of this exercise, i decided to drop them out . This could be forwarded to the responsible teams to ensure they can correct how this information gets captured.

In [10]:
display(df.groupby("Plate_ID").count())

Plate_ID count 62417JM 711 PGQ1276 14 2314816 3 CEG8498 8 GNG1357 4 T521509C 2 P43ECM 5 DWP1806 15 29889BB 3 FYW7780 4 FXB5422 12 C53EJT 13 1AX9819 2 60930MC 321 XX421E 48 GGF4120 6 GWJ5706 3 EPJ7840 2 88654JV 7 86445MB 265 WGX50A 6 FDF2169 1 FAL9284 27 T526896C 19 87399JV 119 23626MA 36 FZF6814 18 66081MB 33 90783 53 GMK5594 3 GHF9893 4 23713MB 149 22240MD 14 AIPT57 1 97645JZ 12 YART033 40 E16DVX 17 GDV1341 25 DBC9461 1 EDK9892 32 GPP5067 3 58443JN 14 37207PC 98 17744MD 926 CF6S629 17 FAM3010 7 GPY1724 1 EEL7193 4 40920JV 285 GHP2172 2 GGH5632 5 89621ML 1 AR491A 11 GPR3557 13 AZE9207 5 JLK7542 2 92565JE 296 GKX3677 14 CVDW11 1 84579JN 4 42640ME 60 FPE3496 28 1Y19D 3 X78ESE 2 GKN5142 2 AP123G 70 BWJ4755 4 75256JX 106 83903MC 145 GCA4556 5 74603MB 19 GCV3955 11 GND7107 4 GGA3301 4 FSX5453 9 29862JV 148 XCDT27 16 33022JB 764 GXV6161 8 FXM3101 1 GSM5335 14 GJH8732 4 GUR5329 2 GWB9197 2 BTV7253 2 96437MA 27 S18ANT 9 21891MC 70 ECC8031 4 2270377 38 BLE5281 4 GPV3679 1 12133JY 232 42020JX 34 GEL1042 5 GJT2433 38 62500JM 75 JKA6662 12 67359MB 61 FGP1122 6 GMU3358 19 GVV7520 10 EML2164 15 CYN8980 4 GFF3314 17 32414MD 57 EJA9662 35 FXB7476 4 42220JZ 26 JMJ5713 8 GWR4723 2 GCM3037 6 18866JZ 46 57315MG 171 C46CYX 1 FAC264 2 59460MD 32 86785MA 54 AG28478 3 EPB6731 5 79390JZ 165 DWM7801 2 GKM8040 4 88403JU 257 T632425C 2 GKJ3772 4 T653265C 5 AH798Y 2 CEB5140 11 GKW8679 2 GJ47OT 1 SODASRV3 40 FYL1120 8 99130ME 37 1AFVU6 1 GFE2524 5 80622JY 77 AN438H 6 824D8 1 43677MD 4 GLZ4866 9 DPJ8751 1 9AHLM6 2 GSY7332 2 BDVU74 9 GGT3490 1 ESX8214 6 DRY4909 4 DLZ8073 2 60967MA 21 CPZ6343 7 JLA8809 19 21855PC 10 72180MD 74 GHR1298 7 88946MA 3 FZB4088 14 UJH67G 4 31590ME 152 41921JZ 104 FPE3784 6 57254JZ 18 PRD657 1 MAY55W 1 GCK2215 1 GKJ5532 2 GDN8440 10 GLR5014 2 562PX4 19 EWX9913 2 FLM3541 3 GSR7204 1 73630MD 17 A46BXY 4 88046JA 82 29866JV 67 GEP3204 8 69779JU 23 2219442 53 2336794 34 N62ECP 5 EAC4076 10 25319BB 11 EVM2038 4 25570LV 1 ACS8987 1 2SX067 25 GNH1620 12 EXV9925 2 ESW2400 4 DUJ4556 7 GUP8778 3 29068JY 31 GJH8132 3 DFN2149 5 98741PA 6 GBL1705 4 84321JY 27 GCU3100 17 FGY1097 3 GJM2309 5 71661KA 8 EMA6556 1 4GEA036 6 86352JZ 73 GGT3878 8 GTE6608 1 GNK1338 3 54919MA 1 30901JZ 40 24996ME 39 DD39542 1 P87ETP 11 20983ME 183 92429JY 261 94837JZ 55 C72CLU 1 FZA5037 16 784CKO 1 DPD9308 6 XCX1911 3 GRE3342 26 FXW3486 3 GKZ8450 9 GPV6650 14 4BCY65 8 GHW4938 9 XAMT58 43 L82EAG 1 B5760165 28 HML4085 1 757KX8 6 GRF2371 5 ZZR5457 1 DVY9299 10 T87DWH 15 EGE9399 1 ERK3680 2 82935ME 193 9004DA 3 XZ477U 21 ALX9056 1 GMF2780 12 GLH223 4 24643MD 22 GLT6530 17 PSNGTHRU 1 GFE9584 4 FGY1922 8 70581JT 11 FFS8916 1 GGF5778 3 M957TK 16 FWF1907 5 T636385C 11 BZD6089 31 FYB1193 1 EKF8411 2 GUL5746 9 33385MB 4 T508977C 17 SIRLANS 2 FMV9676 3 CPL8883 3 61829JY 6 AVM4697 1 GRC8903 12 93714JY 8 FWF3090 14 FMG1564 3 2304069 33 FLW5910 9 T638989C 4 H19BJT 4 39034MG 8 GVH3205 2 GJC4514 7 AU516X 2 61208JV 61 U51ECP 2 GTB6090 6 CUH9869 3 Z35EPP 2 GFV9611 7 T417584C 7 GRZ7840 21 GNX8029 7 53143JJ 80 GGT6009 2 FMT1686 6 FWL9371 5 78206MB 52 40271MG 5 EXS5740 1 X64AHK 2 FHY5051 6 FXM4289 13 BYL8950 3 FTF7545 18 82994AS 2 85143JU 78 50248A 11 CMU7665 5 EYF1086 1 GSC3877 6 GKP9900 10 GYW1471 7 T633432C 3 GGY7988 5 GNX4552 3 CUY2647 1 DEP9275 1 GUY2498 12 NYP5449 2 96526JS 17 135UZL 5 FRS9586 1 876UWS 2 T634672C 5 48160MG 13 BADDONE 1 17094MD 143 HRE3763 2 GNS3160 31 GLH8482 7 39663JR 20 JHM1121 1 EJR7152 1 GDB2241 4 52279IJ 1 ATE2433 2 GTS5229 11 FYA3286 6 GTT7318 20 12391MC 181 61416MB 193 GHT4777 3 51814JS 64 V82FFN 11 GRB7084 5 25339MB 38 FXE3686 5 938XPT 1 GFN7806 14 CCW6110 1 59279JA 46 DDM5119 1 FSR0335 1 DGG7070 19 XR112K 68 GUG8372 24 GTC1083 8 ECON101 5 GJC8350 1 1MP847 1 GAT9322 9 GLV6691 2 FPZ5047 9 GFD8363 14 GCP9549 6 AR335A 218 2257367 17 AP878L 7 XALJ35 19 EED1755 6 AE39151 10 DYN9645 4 2904DLZ 4 GSR3150 7 P91FHC 6 340ARQ 2 DWL3574 1 FYM7108 47 GDT5099 3 41755JX 3 GSV8599 14 FYD3604 11 GLL2019 105 21240BB 6 GKE2431 16 GBK5362 13 G76DBU 10 BN02465 1 GEW9409 9 DVA3602 38 6

From the above tables, we can see that there have vehicles who have been tagged multiple times within a give year. Obviously the 711 occurences looks a bit fishy as the data is for a year and there are only 365 days in a year.

lets check the violation dimensions

In [13]:
display(df.select("Violation_Code","Violation_Legal_Code","Law_Section","Violation_Post_Code","Violation_Description","No_Standing_or_Stopping_Violation","Hydrant_Violation","Double_Parking_Violation"))

Violation_Code Violation_Legal_Code Law_Section Violation_Post_Code Violation_Description No_Standing_or_Stopping_Violation Hydrant_Violation Double_Parking_Violation 21 null 408 A 77 21-No Parking (street clean) null null null 14 null 408 CC3 14-No Standing null null null 46 null 408 J 32 46A-Double Parking (Non-COM) null null null 19 null 408 01 4 19-No Stand (bus stop) null null null 19 null 408 19 7 19-No Stand (bus stop) null null null 21 null 408 C 32 21-No Parking (street clean) null null null 21 null 408 10 3 21-No Parking (street clean) null null null 71 null 408 A 42 71A-Insp Sticker Expired (NYS) null null null 69 null 408 093 69-Failure to Disp Muni Recpt null null null 71 null 408 N 42 71A-Insp Sticker Expired (NYS) null null null 71 null 408 N 42 71A-Insp Sticker Expired (NYS) null null null 20 null 408 17 4 20A-No Parking (Non-COM) null null null 20 null 408 53 1 20A-No Parking (Non-COM) null null null 16 null 408 59 3 16-No Std (Com Veh) Com Plate null null null 85 null 408 G 42 85-Storage-3 hour Commercial null null null 14 null 408 04 6 14-No Standing null null null 21 null 408 J 32 21-No Parking (street clean) null null null 46 null 408 null null null null null 71 null 408 002 71A-Insp Sticker Expired (NYS) null null null 37 null 408 34 2 37-Expired Muni Meter null null null 70 null 408 null null null null null 42 null 408 112 42-Exp. Muni-Mtr (Com. Mtr. Z) null null null 46 null 408 E-AM 46B-Double Parking (Com-100Ft) null null null 20 null 408 CC1 20A-No Parking (Non-COM) null null null 20 null 408 CC1 20A-No Parking (Non-COM) null null null 38 null 408 09 3 38-Failure to Display Muni Rec null null null 20 null 408 J 99 20A-No Parking (Non-COM) null null null 20 null 408 C 32 20A-No Parking (Non-COM) null null null 66 null 408 null null null null null 37 null 408 48 2 37-Expired Muni Meter null null null 46 null 408 28 1 46B-Double Parking (Com-100Ft) null null null 46 null 408 null null null null null 21 null 408 14 3 21-No Parking (street clean) null null null 37 null 408 03 4 37-Expired Muni Meter null null null 21 null 408 55 2 21-No Parking (street clean) null null null 46 null 408 01 3 46B-Double Parking (Com-100Ft) null null null 89 null 408 10 6 89-No Stand Exc Com Plate null null null 38 null 408 36 4 38-Failure to Display Muni Rec null null null 21 null 408 38 4 21-No Parking (street clean) null null null 71 null 408 26 4 71A-Insp Sticker Expired (NYS) null null null 71 null 408 26 4 71A-Insp Sticker Expired (NYS) null null null 37 null 408 15 3 37-Expired Muni Meter null null null 21 null 408 08 4 21-No Parking (street clean) null null null 38 null 408 A 77 38-Failure to Display Muni Rec null null null 14 null 408 17 7 14-No Standing null null null 21 null 408 14 3 21-No Parking (street clean) null null null 21 null 408 14 3 21-No Parking (street clean) null null null 71 null 408 14 3 71A-Insp Sticker Expired (NYS) null null null 38 null 408 03-P 38-Failure to Display Muni Rec null null null 69 null 408 22 6 69-Failure to Disp Muni Recpt null null null 14 null 408 I 42 14-No Standing null null null 14 null 408 I 42 14-No Standing null null null 42 null 408 99 6 42-Exp. Muni-Mtr (Com. Mtr. Z) null null null 71 null 408 04A 71A-Insp Sticker Expired (NYS) null null null 71 null 408 04A 71A-Insp Sticker Expired (NYS) null null null 37 null 408 06 3 37-Expired Muni Meter null null null 14 null 408 99 6 14-No Standing null null null 20 null 408 56 7 20-No Parking (Com Plate) null null null 21 null 408 O 41 21-No Parking (street clean) null null null 38 null 408 11P 38-Failure to Display Muni Rec null null null 21 null 408 31 4 21-No Parking (street clean) null null null 21 null 408 Q 21 21-No Parking (street clean) null null null 21 null 408 12A 21-No Parking (street clean) null null null 37 null 408 C 77 37-Expired Muni Meter null null null 38 null 408 null null null null null 38 null 408 56 3 38-Failure to Display Muni Rec null null null 37 null 408 41 2 37-Expired Muni Meter null null null 71 null 

we already know from our counts earlier that the data for No_Standing_or_Stopping , Hydrant and Double_Parking violation are not available. We could possibly parse the description of the violation to extract  them.

In [15]:
display(df.groupby("Violation_Code","Law_Section","Violation_Description","Violation_Post_Code").count())

Violation_Code Law_Section Violation_Description Violation_Post_Code count 21 408 21-No Parking (street clean) 66 3631 40 408 40-Fire Hydrant C 77 11243 21 408 21-No Parking (street clean) E 42 5047 20 408 20A-No Parking (Non-COM) 20-A 91 16 408 16-No Std (Com Veh) Com Plate 14 7 615 71 408 71A-Insp Sticker Expired (NYS) I 42 4447 20 408 20-No Parking (Com Plate) 95 1 2031 20 408 20-No Parking (Com Plate) 092 135 38 408 38-Failure to Display Muni Rec 02P 1429 70 408 70A-Reg. Sticker Expired (NYS) 02 3 845 31 408 31-No Stand (Com. Mtr. Zone) 58 6 148 20 408 20-No Parking (Com Plate) 55 1 1122 70 408 70B-Impropr Dsply of Reg (NYS) B 42 114 70 408 70B-Impropr Dsply of Reg (NYS) 04-A 68 48 408 48-Bike Lane 70 1 222 68 408 68-Not Pkg. Comp. w Psted Sign C 77 85 46 408 46B-Double Parking (Com-100Ft) J 99 254 37 408 37-Expired Muni Meter 210J 1476 51 408 51-Sidewalk 01 4 445 70 408 70-Reg. Sticker Missing (NYS) 04 6 11 61 408 61-Wrong Way 06 6 21 21 408 21-No Parking (street clean) 11-A 712 21 408 21-No Parking (street clean) 14-A 422 46 408 46A-Double Parking (Non-COM) C1 3 1237 19 408 19-No Stand (bus stop) 03-A 164 16 408 16A-No Std (Com Veh) Non-COM 132 231 50 408 50-Crosswalk T 41 177 40 408 40-Fire Hydrant G 10 198 46 408 46B-Double Parking (Com-100Ft) B 10 18 16 408 16A-No Std (Com Veh) Non-COM 50-A 66 46 408 46A-Double Parking (Non-COM) 60 7 158 49 408 49-Excavation (obstruct traff) 58 7 13 85 408 85-Storage-3 hour Commercial I 31 179 98 408 98-Obstructing Driveway 03 3 40 48 408 48-Bike Lane D1 3 15 21 408 21-No Parking (street clean) 07 7 1000 46 408 46A-Double Parking (Non-COM) 11P 322 85 408 85-Storage-3 hour Commercial E 77 352 74 408 74A-Improperly Displayed Plate L 32 33 18 408 18-No Stand (bus lane) N 99 65 84 408 84-Platform lifts in low posit 02 4 85 19 408 19-No Stand (bus stop) 48 3 72 61 408 61-Wrong Way A 22 20 46 408 46B-Double Parking (Com-100Ft) 53 1 183 83 408 83-Improper Registration CC5 18 20 408 20A-No Parking (Non-COM) 08-P 260 50 408 50-Crosswalk 015 50 82 408 82-Unaltered Commerc Vehicle 68 6 27 89 408 89-No Stand Exc Com Plate 67 6 103 38 408 38-Failure to Display Muni Rec 23 3 277 14 408 14-No Standing 56 6 98 40 408 40-Fire Hydrant 02-A 211 38 408 38-Failure to Display Muni Rec 092 630 20 408 20A-No Parking (Non-COM) 93 1 12 20 408 20A-No Parking (Non-COM) 45-P 82 38 408 38-Failure to Display Muni Rec 75 6 48 23 408 23-No Parking (taxi stand) 19A 20 84 408 84-Platform lifts in low posit 04-P 9 24 408 24-No Parking (exc auth veh) CC7 26 24 408 24-No Parking (exc auth veh) 17 3 122 71 408 71-Insp. Sticker Missing (NYS 12P 22 52 408 52-Intersection A 42 2 67 408 67-Blocking Ped. Ramp C 62 45 74 408 74-Missing Display Plate F 42 254 51 408 51-Sidewalk 07 2 26 20 408 20A-No Parking (Non-COM) S 42 72 24 408 24-No Parking (exc auth veh) 16 3 64 46 408 46B-Double Parking (Com-100Ft) 59 2 162 70 408 70-Reg. Sticker Missing (NYS) J 42 92 19 408 19-No Stand (bus stop) 014 108 51 408 51-Sidewalk 57-P 41 37 408 37-Expired Muni Meter 21-P 212 98 408 98-Obstructing Driveway 25 4 20 74 408 74B-Covered Plate T 31 27 71 408 71A-Insp Sticker Expired (NYS) 80 6 14 83 408 83-Improper Registration 03A 35 31 408 31-No Stand (Com. Mtr. Zone) 84 6 55 16 408 16A-No Std (Com Veh) Non-COM 18 6 101 46 408 46-Double Parking (Com Plate) 09 7 40 83 408 83-Improper Registration G-PM 5 72 408 72A-Insp Stkr Counterfeit 52-A 22 10 408 10-No Stopping 01-A 28 20 408 20A-No Parking (Non-COM) 37-A 93 4 408 04A-Downtown Bus Area, Non-Bus Y 99 32 16 408 16A-No Std (Com Veh) Non-COM 53-A 77 53 408 53-Safety Zone 62 2 7 20 408 20-No Parking (Com Plate) 93 1 24 40 408 40-Fire Hydrant 027 32 84 408 84-Platform lifts in low posit 21P 14 39 408 39-Overtime PKG-Time Limt Post 52 2 10 74 408 74A-Improperly Displayed Plate C-AM 1 74 408 74-Missing Display Plate S1 9 1 84 408 84-Platform lifts in low posit C 10 19 71 408 71A-Insp Sticker Expired (NYS) 027 36 69 408 69-Failure to Disp Muni Recpt 23 7 32 73 408 73-Reg Stkr Mutilated 15 4 29 70 408 70-R

from the data above we can see that there is a clear 1-1 relationship between violation code,violation post code and violation description. I am not clear as to what the violation post code means. But, i can notice that it could be different for the same violation code.

In [17]:
display(df.select("Violation_Code","Issue_Date","Issuer_Code","Plate_ID","Violation_Legal_Code","Law_Section","Violation_Precinct","Violation_Post_Code","Violation_Description"))

Violation_Code Issue_Date Issuer_Code Plate_ID Violation_Legal_Code Law_Section Violation_Precinct Violation_Post_Code Violation_Description 21 10/01/2014 345454 EPC5238 null 408 7 A 77 21-No Parking (street clean) 14 03/06/2015 333386 5298MD null 408 25 CC3 14-No Standing 46 07/28/2014 331845 FYW2775 null 408 72 J 32 46A-Double Parking (Non-COM) 19 04/13/2015 355669 GWE1987 null 408 102 01 4 19-No Stand (bus stop) 19 05/19/2015 341248 T671196C null 408 28 19 7 19-No Stand (bus stop) 21 11/20/2014 357104 JJF6834 null 408 67 C 32 21-No Parking (street clean) 21 08/01/2014 355251 FYP8263 null 408 79 10 3 21-No Parking (street clean) 71 07/19/2014 353083 FZL7450 null 408 103 A 42 71A-Insp Sticker Expired (NYS) 69 09/23/2014 346381 86318MA null 408 5 093 69-Failure to Disp Muni Recpt 71 01/20/2015 361082 GKX8095 null 408 113 N 42 71A-Insp Sticker Expired (NYS) 71 01/20/2015 361082 GKX8095 null 408 113 N 42 71A-Insp Sticker Expired (NYS) 20 02/17/2015 359625 CXT8949 null 408 109 17 4 20A-No Parking (Non-COM) 20 05/05/2015 361097 GSR4914 null 408 5 53 1 20A-No Parking (Non-COM) 16 10/21/2014 361332 XBVP75 null 408 78 59 3 16-No Std (Com Veh) Com Plate 85 08/21/2014 354200 AP515N null 408 110 G 42 85-Storage-3 hour Commercial 14 08/07/2014 345058 16238TC null 408 14 04 6 14-No Standing 21 08/26/2014 332021 122YEZ null 408 72 J 32 21-No Parking (street clean) 46 06/23/2014 942227 65920JW null 408 19 null null 71 09/24/2014 361338 GBU2135 null 408 9 002 71A-Insp Sticker Expired (NYS) 37 10/04/2014 361051 6AJWV5 null 408 48 34 2 37-Expired Muni Meter 70 01/21/2015 919749 GDL7305 null 408 25 null null 42 03/26/2015 357684 58359MD null 408 14 112 42-Exp. Muni-Mtr (Com. Mtr. Z) 46 10/06/2014 358655 60304MD null 408 114 E-AM 46B-Double Parking (Com-100Ft) 20 01/24/2015 354084 GJP9663 null 408 30 CC1 20A-No Parking (Non-COM) 20 01/24/2015 354084 GJP9663 null 408 30 CC1 20A-No Parking (Non-COM) 38 03/11/2015 358894 G78DLL null 408 79 09 3 38-Failure to Display Muni Rec 20 11/19/2014 353194 GKP6078 null 408 7 J 99 20A-No Parking (Non-COM) 20 02/24/2015 333002 GAR6688 null 408 70 C 32 20A-No Parking (Non-COM) 66 08/20/2014 527637 1545236 null 408 75 null null 37 11/14/2014 356985 M80ANH null 408 48 48 2 37-Expired Muni Meter 46 12/23/2014 359578 25X944 null 408 7 28 1 46B-Double Parking (Com-100Ft) 46 09/15/2014 952479 36913MA null 408 13 null null 21 06/09/2015 355674 GKB7155 null 408 90 14 3 21-No Parking (street clean) 37 09/05/2014 361042 CJP1169 null 408 103 03 4 37-Expired Muni Meter 21 12/18/2014 361369 EYB8055 null 408 52 55 2 21-No Parking (street clean) 46 04/07/2015 357114 96313JE null 408 84 01 3 46B-Double Parking (Com-100Ft) 89 09/11/2014 359588 60154JZ null 408 18 10 6 89-No Stand Exc Com Plate 38 08/12/2014 355608 FFE3632 null 408 108 36 4 38-Failure to Display Muni Rec 21 07/14/2014 357710 GSM7092 null 408 108 38 4 21-No Parking (street clean) 71 01/21/2015 361287 44189JW null 408 114 26 4 71A-Insp Sticker Expired (NYS) 71 01/21/2015 361287 44189JW null 408 114 26 4 71A-Insp Sticker Expired (NYS) 37 09/09/2014 358930 G98BVF null 408 90 15 3 37-Expired Muni Meter 21 05/26/2015 361861 33451MB null 408 107 08 4 21-No Parking (street clean) 38 11/29/2014 355156 JFS9782 null 408 19 A 77 38-Failure to Display Muni Rec 14 07/31/2014 345536 18846MC null 408 19 17 7 14-No Standing 21 01/12/2015 345228 W73DRS null 408 90 14 3 21-No Parking (street clean) 21 01/12/2015 345228 W73DRS null 408 90 14 3 21-No Parking (street clean) 71 12/11/2014 358649 CME3770 null 408 90 14 3 71A-Insp Sticker Expired (NYS) 38 08/25/2014 356557 GEB3048 null 408 122 03-P 38-Failure to Display Muni Rec 69 11/24/2014 359305 55065MB null 408 14 22 6 69-Failure to Disp Muni Recpt 14 01/03/2015 356555 GLY1430 null 408 102 I 42 14-No Standing 14 01/03/2015 356555 GLY1430 null 408 102 I 42 14-No Standing 42 07/28/2014 355231 75455MC null 408 18 99 6 42-Exp. Muni-Mtr (Com. Mtr. Z) 71 01/21/2015 361790 EDT9406 null 408 43 04A 71A-Insp Sticker Expired (NYS) 71 01/21/201

by looking at the above data, i come to a conclusion that though violation_code and violation_post_code seem to having a 1-1 with the description and form a natural key, i don't think that is the case here, it seems to me to be more factual than a dimension. for now i will keep it out of the dimension table, just to sure.

Now lets take a look at the Issuer of the ticket data.

In [20]:
display(df.groupby("Issuer_Code","Issuer_Precinct","Issuer_Command","Issuer_Squad","Issuing_Agency").count())

Issuer_Code Issuer_Precinct Issuer_Command Issuer_Squad Issuing_Agency count 345221 19 T103 B T 6171 350285 18 T106 B T 3559 340060 88 T730 C T 23 355209 20 T103 A T 4000 356550 10 T800 B2 T 451 359323 60 T302 K T 63 358658 17 T106 J T 914 361783 114 T401 P T 1360 361084 94 T301 S T 546 357737 19 T103 O T 4897 361032 68 T302 K T 834 946304 83 0083 0000 P 35 351964 81 T301 Y T 155 356214 24 T103 O T 693 356975 44 T201 M T 1759 358941 114 T401 A T 710 357695 68 T302 C T 186 358579 1 T105 H T 3418 345178 14 T108 A T 148 357070 109 T401 J T 1673 361281 68 T302 N T 129 583618 0 2W09 0000 S 17 358949 67 T302 P T 100 931249 49 0049 0000 P 2835 356182 66 T302 B T 206 361110 5 T101 A T 507 357808 70 T302 K T 494 354002 111 T401 G T 76 361354 107 T402 L T 1224 356329 114 T401 O T 701 330198 6 T101 M T 3328 356491 34 T103 Q T 261 361062 19 T103 Q T 1410 669521 0 KS07 0000 S 1156 343920 77 T301 G T 702 355179 7 T101 B T 388 356537 106 T402 P T 1051 355594 52 T201 N T 495 359325 23 T103 O T 227 356345 77 T301 S T 128 915975 44 0044 0000 P 41 340095 9 T800 A2 T 321 358927 30 T103 N T 46 346199 73 T301 Y T 72 354064 103 T803 A T 316 358223 19 T103 R T 804 361329 70 T302 R T 392 358846 68 T302 R T 317 361835 62 T302 R T 315 358208 62 T302 C T 135 534941 0 0PI6 0000 S 4 945551 94 0094 0000 P 39 945824 48 0048 0000 P 366 362267 33 T103 Q T 45 950329 67 0067 0000 P 46 361311 46 T201 H T 91 361808 17 T102 D T 15 355437 18 T106 L T 1027 946747 69 0069 0000 P 56 948426 102 0102 0000 P 67 928729 104 0104 0000 P 159 943628 6 0006 0000 X 2 356236 109 T401 L T 155 356965 111 T401 H T 61 359313 41 T202 F T 193 358927 33 T103 N T 179 356314 34 T103 H T 134 356243 25 T103 Q T 50 358895 100 T402 C T 56 361024 10 T102 L T 202 345494 78 T730 F T 229 355157 13 T102 D T 462 906743 60 0060 0000 P 63 955895 42 0042 0000 P 107 350410 84 T802 B T 398 955079 165 0165 0000 P 63 361283 17 T102 F T 284 361329 63 T302 R T 210 361031 63 T302 P T 388 353094 17 T102 D T 31 361824 60 T302 O T 156 356300 26 T103 G T 148 931890 62 0062 0000 P 112 357093 78 T302 A T 254 362171 61 T302 G T 114 937838 69 0069 0000 P 127 611164 0 QE10 0000 S 17 902687 110 0110 0000 P 142 941019 73 0073 0000 P 50 952067 46 0046 0000 P 20 347447 66 T802 B T 70 908338 977 0977 0000 X 128 331652 0 T108 J T 2 349528 81 T301 Y T 60 358605 46 T202 F T 18 343926 71 T802 B T 43 901787 0 0MTS 0000 P 82 362193 5 T101 E T 13 932790 43 0043 0000 X 23 343942 113 T402 M T 471 956295 71 0071 0000 P 134 956321 47 0047 0000 X 21 331335 13 T108 I T 13 956007 101 0101 0000 P 29 934560 24 0024 0000 X 175 361290 52 T201 X T 211 916697 43 0043 0000 P 49 951201 23 0023 0000 P 22 361347 28 T103 Q T 48 361131 72 T302 R T 25 936215 50 0050 0000 P 77 956268 103 0103 0000 P 41 362170 67 T302 G T 60 355253 24 T103 D T 175 948217 68 0068 0000 P 93 357115 73 T301 E T 56 160232 3 0003 0000 P 32 357794 72 T302 I T 21 941070 67 0067 0000 P 1 601532 0 MN06 0000 X 42 929534 73 0073 0000 P 29 948720 14 0014 0000 P 7 953216 120 0120 0000 P 6 352089 17 T400 A T 67 906732 23 0023 0000 X 11 946936 62 0062 0000 P 61 346004 19 T108 M T 3 355249 110 T401 P T 16 357368 73 T301 O T 216 361290 43 T201 R T 17 357799 71 T302 O T 100 953092 49 0049 0000 X 5 333163 102 T402 Y T 40 351877 41 T202 J T 38 902520 32 0032 0000 X 4 358929 26 T103 R T 19 958000 0 KS14 0000 S 223 955634 162 0162 0000 P 69 350442 78 T302 P T 173 330001 20 T503 B1 T 72 939256 25 0025 0000 X 7 927716 108 0108 0000 P 87 925634 109 0109 0000 P 67 945936 43 0043 0000 X 8 930597 111 0111 0000 P 111 219 835 0835 0000 U 8 347599 84 T301 P T 18 600133 982 0982 0000 X 48 956835 7 0007 0000 X 4 950100 9 0009 0000 P 1 948862 68 0068 0000 P 105 956322 171 0171 0000 P 6 949322 13 0013 0000 P 32 950517 109 0109 0000 P 1 955197 70 0070 0000 X 9 362244 104 T401 T T 5 937976 73 0073 0000 X 1 0 0 MN11 0000 S 10 919064 105 0105 0000 P 74 943952 806 0806 0000 P 6 358937 71 T302 Y T 20 956260 79 0079 0000 P 50 949094 67 0067 0000 X 3 937545 102 0102 0000 P 36 957075 13 0013 0000 P 46 515323 0 

from the groupby above, i could clearly reveal a relation between issuer code and the issuer_squad and his agency. But ,we can also , that for a few records the data has not been captured accurately. For the purpose of this exercise, i will be again dropping all records that have issuer_code 0. I can forward these records also to the ingest team so they can correct the data acquisition.

In [22]:
display(df.where(df["Issuer_Code"]=="330581").select("Issuer_Code","Violation_Precinct","Issue_Date","Issuer_Precinct","Issuer_Command","Issuer_Squad","Issuing_Agency"))

Issuer_Code Violation_Precinct Issue_Date Issuer_Precinct Issuer_Command Issuer_Squad Issuing_Agency 330581 14 06/10/2015 14 T501 B T 330581 14 01/31/2015 14 T501 B T 330581 14 06/04/2015 14 T501 B T 330581 14 03/21/2015 14 T501 B T 330581 18 10/05/2014 18 T501 B T 330581 17 05/03/2015 17 T501 B T 330581 17 05/03/2015 17 T501 B T 330581 10 11/16/2014 10 T501 B T 330581 18 01/24/2015 18 T501 B T 330581 18 01/24/2015 18 T501 B T 330581 18 05/04/2015 18 T501 B T 330581 10 07/24/2014 10 T501 B T 330581 13 01/25/2015 13 T501 B T 330581 13 01/25/2015 13 T501 B T 330581 14 11/22/2014 14 T501 B T 330581 10 05/20/2015 10 T501 B T 330581 13 06/11/2015 13 T501 B T 330581 20 05/11/2015 20 T501 B T 330581 17 06/08/2015 17 T501 B T 330581 6 08/07/2014 6 T501 B T 330581 17 03/16/2015 17 T501 B T 330581 24 10/14/2014 24 T501 B T 330581 7 09/09/2014 7 T501 B T 330581 13 07/16/2014 13 T501 B T 330581 14 06/12/2015 14 T501 B T 330581 6 05/03/2015 6 T501 B T 330581 6 05/10/2015 6 T501 B T 330581 14 07/28/2014 14 T501 B T 330581 6 04/22/2015 6 T501 B T 330581 10 02/10/2015 10 T501 B T 330581 17 07/22/2014 17 T501 B T 330581 19 10/08/2014 19 T501 B T 330581 13 03/16/2015 13 T501 B T 330581 10 10/29/2014 10 T501 B T 330581 14 05/16/2015 14 T501 B T 330581 14 05/03/2015 14 T501 B T 330581 6 05/27/2015 6 T501 B T 330581 7 01/11/2015 7 T501 B T 330581 7 01/11/2015 7 T501 B T 330581 13 03/03/2015 13 T501 B T 330581 13 11/13/2014 13 T501 B T 330581 6 04/22/2015 6 T501 B T 330581 13 03/07/2015 13 T501 B T 330581 18 07/06/2014 18 T501 B T 330581 6 07/09/2014 6 T501 B T 330581 17 01/14/2015 17 T501 B T 330581 17 01/14/2015 17 T501 B T 330581 17 09/25/2014 17 T501 B T 330581 28 06/24/2015 28 T501 B T 330581 18 05/31/2015 18 T501 B T 330581 14 02/24/2015 14 T501 B T 330581 13 01/31/2015 13 T501 B T 330581 17 06/27/2015 17 T501 B T 330581 19 04/14/2015 19 T501 B T 330581 24 03/20/2015 24 T501 B T 330581 14 01/29/2015 14 T501 B T 330581 17 12/29/2014 17 T501 B T 330581 17 12/29/2014 17 T501 B T 330581 10 01/24/2015 10 T501 B T 330581 10 01/24/2015 10 T501 B T 330581 19 09/23/2014 19 T501 B T 330581 18 10/05/2014 18 T501 B T 330581 10 06/14/2015 10 T501 B T 330581 6 05/27/2015 6 T501 B T 330581 115 12/16/2014 115 T501 B T 330581 6 05/17/2015 6 T501 B T 330581 7 09/07/2014 7 T501 B T 330581 6 04/16/2015 6 T501 B T 330581 17 02/23/2015 17 T501 B T 330581 6 04/22/2015 6 T501 B T 330581 13 03/16/2015 13 T501 B T 330581 17 02/21/2015 17 T501 B T 330581 13 02/11/2015 13 T501 B T 330581 13 01/31/2015 13 T501 B T 330581 14 12/29/2014 14 T501 B T 330581 14 12/29/2014 14 T501 B T 330581 10 09/28/2014 10 T501 B T 330581 18 08/15/2014 18 T501 B T 330581 10 09/28/2014 10 T501 B T 330581 14 05/16/2015 14 T501 B T 330581 13 05/11/2015 13 T501 B T 330581 14 05/02/2015 14 T501 B T 330581 71 06/26/2015 71 T501 B T 330581 6 04/22/2015 6 T501 B T 330581 6 04/16/2015 6 T501 B T 330581 6 09/08/2014 6 T501 B T 330581 13 12/29/2014 13 T501 B T 330581 13 12/29/2014 13 T501 B T 330581 14 02/15/2015 14 T501 B T 330581 17 09/03/2014 17 T501 B T 330581 18 05/12/2015 18 T501 B T 330581 10 02/09/2015 10 T501 B T 330581 14 06/05/2015 14 T501 B T 330581 13 03/03/2015 13 T501 B T 330581 13 10/13/2014 13 T501 B T 330581 1 06/27/2014 1 T501 B T 330581 13 01/24/2015 13 T501 B T 330581 13 01/24/2015 13 T501 B T 330581 17 02/21/2015 17 T501 B T 330581 10 09/28/2014 10 T501 B T 330581 17 03/07/2015 17 T501 B T 330581 17 01/24/2015 17 T501 B T 330581 17 01/24/2015 17 T501 B T 330581 19 10/20/2014 19 T501 B T 330581 17 10/15/2014 17 T501 B T 330581 1 04/16/2015 1 T501 B T 330581 14 01/22/2015 14 T501 B T 330581 14 01/22/2015 14 T501 B T 330581 18 08/09/2014 18 T501 B T 330581 17 10/06/2014 17 T501 B T 330581 13 09/21/2014 13 T501 B T 330581 6 04/16/2015 6 T501 B T 330581 13 01/20/2015 13 T501 B T 330581 13 01/20/2015 13 T501 B T 330581 10 08/10/2014 10 T501 B T 330581 13 01/23/2015 13 T501 B T 330581 13 01/23/2015 13 T501 B T 330581 10 06/30/2014 10 T501 B T 330581 10 01/11/2015 10 T501 B T 330581 10 01

In [23]:
display(df.where(df["Issuer_Code"]=="347447").select("Issuer_Code","Violation_Precinct","Issue_Date","Issuer_Precinct","Issuer_Command","Issuer_Squad","Issuing_Agency"))

Issuer_Code Violation_Precinct Issue_Date Issuer_Precinct Issuer_Command Issuer_Squad Issuing_Agency 347447 94 09/30/2014 94 T802 B T 347447 88 11/05/2014 88 T802 B T 347447 110 08/06/2014 110 T803 B T 347447 108 08/14/2014 108 T803 B T 347447 76 10/28/2014 76 T802 B T 347447 76 10/28/2014 76 T802 B T 347447 61 09/26/2014 61 T802 B T 347447 83 09/30/2014 83 T802 B T 347447 84 11/04/2014 84 T802 B T 347447 103 08/27/2014 103 T803 B T 347447 102 07/24/2014 102 T803 B T 347447 103 08/22/2014 103 T803 B T 347447 102 08/09/2014 102 T803 B T 347447 105 08/05/2014 105 T803 B T 347447 94 10/04/2014 94 T802 B T 347447 114 06/29/2014 114 T803 B T 347447 112 07/23/2014 112 T803 B T 347447 70 09/17/2014 70 T802 B T 347447 115 07/24/2014 115 T803 B T 347447 109 08/04/2014 109 T803 B T 347447 84 10/23/2014 84 T802 B T 347447 94 09/24/2014 94 T802 B T 347447 107 08/21/2014 107 T803 B T 347447 109 07/11/2014 109 T803 B T 347447 108 08/31/2014 108 T803 B T 347447 108 07/08/2014 108 T803 B T 347447 107 07/10/2014 107 T803 B T 347447 75 10/09/2014 75 T802 B T 347447 110 06/28/2014 110 T803 B T 347447 108 08/31/2014 108 T803 B T 347447 75 10/09/2014 75 T802 B T 347447 108 08/15/2014 108 T803 B T 347447 88 10/15/2014 88 T802 B T 347447 103 07/26/2014 103 T803 B T 347447 108 08/15/2014 108 T803 B T 347447 88 10/30/2014 88 T802 B T 347447 73 09/24/2014 73 T802 B T 347447 108 08/14/2014 108 T803 B T 347447 107 08/20/2014 107 T803 B T 347447 78 10/14/2014 78 T802 B T 347447 111 07/11/2014 111 T803 B T 347447 114 07/30/2014 114 T803 B T 347447 115 07/30/2014 115 T803 B T 347447 109 06/27/2014 109 T803 B T 347447 113 08/21/2014 113 T803 B T 347447 63 09/04/2014 63 T802 B T 347447 94 10/03/2014 94 T802 B T 347447 111 06/27/2014 111 T803 B T 347447 108 06/26/2014 108 T803 B T 347447 109 07/31/2014 109 T803 B T 347447 77 09/23/2014 77 T802 B T 347447 67 10/02/2014 67 T802 B T 347447 94 09/25/2014 94 T802 B T 347447 108 08/15/2014 108 T803 B T 347447 70 09/10/2014 70 T802 B T 347447 94 09/30/2014 94 T802 B T 347447 69 09/05/2014 69 T802 B T 347447 69 09/04/2014 69 T802 B T 347447 104 08/14/2014 104 T803 B T 347447 109 08/19/2014 109 T803 B T 347447 72 10/30/2014 72 T802 B T 347447 78 10/16/2014 78 T802 B T 347447 68 10/29/2014 68 T802 B T 347447 107 08/21/2014 107 T803 B T 347447 106 08/09/2014 106 T803 B T 347447 66 10/22/2014 66 T802 B T 347447 88 10/24/2014 88 T802 B T 347447 77 10/28/2014 77 T802 B T 347447 102 07/26/2014 102 T803 B T 347447 104 08/06/2014 104 T803 B T 347447 105 06/28/2014 105 T803 B T 347447 60 10/05/2014 60 T802 B T 347447 84 10/24/2014 84 T802 B T 347447 72 10/30/2014 72 T802 B T 347447 60 10/04/2014 60 T802 B T 347447 84 10/31/2014 84 T802 B T 347447 108 08/15/2014 108 T803 B T 347447 84 10/25/2014 84 T802 B T 347447 112 08/09/2014 112 T803 B T 347447 112 07/14/2014 112 T803 B T 347447 94 10/23/2014 94 T802 B T 347447 114 07/10/2014 114 T803 B T 347447 61 09/04/2014 61 T802 B T 347447 115 07/09/2014 115 T803 B T 347447 72 09/10/2014 72 T802 B T 347447 108 07/29/2014 108 T803 B T 347447 78 10/31/2014 78 T802 B T 347447 61 09/03/2014 61 T802 B T 347447 112 07/23/2014 112 T803 B T 347447 102 08/09/2014 102 T803 B T 347447 112 07/16/2014 112 T803 B T 347447 111 08/29/2014 111 T803 B T 347447 109 08/19/2014 109 T803 B T 347447 70 09/21/2014 70 T802 B T 347447 67 09/16/2014 67 T802 B T 347447 76 10/28/2014 76 T802 B T 347447 75 09/24/2014 75 T802 B T 347447 114 07/31/2014 114 T803 B T 347447 105 07/15/2014 105 T803 B T 347447 72 10/29/2014 72 T802 B T 347447 108 08/31/2014 108 T803 B T 347447 108 07/30/2014 108 T803 B T 347447 107 08/21/2014 107 T803 B T 347447 84 11/01/2014 84 T802 B T 347447 104 08/14/2014 104 T803 B T 347447 109 07/31/2014 109 T803 B T 347447 66 09/18/2014 66 T802 B T 347447 75 10/28/2014 75 T802 B T 347447 107 08/21/2014 107 T803 B T 347447 108 08/30/2014 108 T803 B T 347447 102 07/27/2014 102 T803 B T 347447 60 09/18/2014 60 T802 B T 347447 108 07/17/2014 108 T803 B T 347447 90 09/11/2014 90 T802 B T 347447 113 07/2

from the above filtered data for one of the issuer_code, it seems that issuer code is tightly aligned with the command, squad and agency. The precint seems to be more of a fact like data of where the violation occured. So issuer_code, issuer_command, issuer_squad and issuing_agency would be the columns for our dimension.Looks like there is a 1-1 relation between issuuer code , issuer command, issuer squad and issuer agency. But, the issuer precinct is changing

Lets look at the location fields. From the earlier count we could see that there was not latitude/longitude data available so I dropped it from my analysis for this exercise. We could may be use the address and query a service like google maps or a GIS system for the latitude and longitude, and enrich the data as it comes in.

In [26]:
display(df.select("Street_Code1","Street_Code2","Street_Code3","Violation_County","Street_Name","House_Number","Intersecting_Street"))

Street_Code1 Street_Code2 Street_Code3 Violation_County Street_Name House_Number Intersecting_Street 20390 29890 31490 NY Essex St 133 null 27790 19550 19570 NY Park Ave 1916 null 8130 5430 5580 K 31st St 184 null 59990 16540 16790 Q Queens Blvd 120-20 null 36090 10410 24690 NY W 116th St 66 null 74230 37980 38030 K Rutland Rd 1013 null 77530 63030 93230 K Skillman St 155 null 49210 0 0 Q Henderson Ave S 30ft W/of 189th St 21690 26990 26790 NY Grand St 199 null 35490 35780 22670 Q Bedell St 137-22 null 35490 35780 22670 Q Bedell St 137-22 null 51190 9140 61090 Q Janet Pl 39-15 null 20090 14010 22890 NY Eldridge St 48 null 5780 24830 43130 K 7th Ave 138 null 10790 19040 18990 Q 50th Ave 97-11 null 13610 34410 34430 NY Broadway 1260 null 5580 9080 9130 K 5th Ave 5022 null 18770 10010 45590 NY E 89 428 null 10110 17190 17210 NY 2nd Ave 166 null 40220 27180 27270 BX Hoffman St 2390 null 19330 10010 29090 NY EAST 119 STR 414 null 18070 0 0 NY E 54th St N 20ft E/of 5th Ave 8140 8590 8690 Q 31st St 35-34 null 27390 36390 36430 NY Old Broadway 94 null 27390 36390 36430 NY Old Broadway 94 null 49130 67030 64830 K Herkimer St 121 null 15010 15510 26440 NY Cherry St 340 null 26630 67030 64830 K Clarkson Ave 324 null 62030 81050 94630 K MONTAUK AVE 651 null 72720 26760 26790 BX Walton Ave 1405 null 16310 0 0 NY Delancey St N 10ft E/of Columbia S 10510 34230 34250 NY 6TH AVE 716 null 89030 60730 57030 K Wallabout St 188 null 64790 18640 18790 Q Sutphin Blvd 94-14 null 0 0 0 BX Bainbridge Ave W 10ft S/of E 208th St 53830 29830 42730 K Joralemon St 210 null 10610 34530 34550 NY 7th Ave 533 null 47990 10240 10390 Q Greenpoint Ave 45-17 null 10540 10590 10790 Q 47th St 48-11 null 6740 8590 8690 Q 21st St 35-34 null 6740 8590 8690 Q 21st St 35-34 null 44930 29230 87630 K Graham Ave 45 null 23940 17040 17130 Q 159th St 84-06 null 10110 19190 19210 NY 2nd Ave 2170 null 18250 25390 27790 NY E 63rd St 41 null 45130 19530 18630 K Grand St 147 null 45130 19530 18630 K Grand St 147 null 78230 19530 18630 K S 2nd St 113 null 39800 40404 40404 R New Dorp Lane Munici null null 34410 10410 13610 NY W 32nd St 11 null 18290 17740 17990 Q 91st Ave 8719 null 18290 17740 17990 Q 91st Ave 8719 null 10510 34810 34830 NY 6th Ave 1301 null 77150 50020 78485 BX White Plains Rd 1354 null 77150 50020 78485 BX White Plains Rd 1354 null 0 40404 40404 K Flatbush Ave Exten 340 null 10910 34750 34770 NY 9th Ave 729 null 10210 18670 18690 NY 3rd Ave 1505 null 57520 19540 19740 Q Nicolls Ave 102-32 null 8320 55520 77150 BX Allerton Ave 677 null 7740 7990 37290 Q 29th St 31-75 null 72320 26670 40520 BX Vyse Ave 1173 null 53920 55720 18520 BX Newbold Ave 2146 null 10110 18590 18610 NY 2nd Ave 1546 null 41930 0 0 K FRANKLIN AV 840 E 6630 5430 5580 K 15th St 205 null 29320 10020 73520 BX E Gun Hill Rd 171 null 27820 31390 31510 Q ALMEDA AVE 54-09 null 57790 0 0 Q Northern Blvd N 45ft E/of 38th St 10010 0 0 NY 1st Ave E 1ft S/of E 75th St 90680 0 0 K W 4th St E 5ft S/of Avenue P 90530 9830 14380 K W 1st St 1543 null 17790 0 0 NY E 40th St N 10ft W/of Madison Av 37370 21190 28890 NY W 181st St 815 null 76950 72820 49020 BX Westchester Ave 1563 null 34650 30550 10810 NY W 44th St 234 null 35190 11710 44990 NY W 71st St 276 null 37030 74230 93830 K E 51st St 145 null 19390 27790 24890 NY E 122nd St 135 null 42490 8640 8740 Q Ditmars Blvd 35-04 null 16110 33910 13310 NY Cornelia St 37 null 35470 14510 15710 NY W 85th St 68 null 29020 52120 21820 BX E Kingsbridge Rd 58 null 36950 11710 13610 NY W 159th St 521 null 25390 17610 17630 NY Madison Ave 136 null 34430 10410 13610 NY W 33 ST 10 null 17590 24890 10210 NY E 30th St 158 null 51790 8190 57790 Q Junction Blvd 32-09 null 18450 10410 25390 NY E 73rd St 4 null 13610 35470 35490 NY Broadway 2350 null 34310 10410 13610 NY W 27th St 11 null 13610 34790 34810 NY Broadway 1657 null 35490 15710 11710 NY W 86 ST 165 null 37462 40404 40404 K E 64th St 6315 null 24890 18670 18690 NY Lexington Ave 1258 null 17450 0 0 NY E 23rd St N 5ft E/of 2

from the above data we can see that, Street_Code1 is related to the actual Street_Name, the Street_Code2 and Street_Code2 are derived from the house number. They are both set to 0 , when there is not house number or a intersecting street value is set. This ideally could be a dimensional data, were based on the address i could populate the street codes and the latitude/longitude for some analytics. I will not use it for my exercise, but I will add these fields to a seperate table, so we can start building on that set. I can include the precinct and meter_number info in this set, so may be we can show the details of a precin